runtime/gc_vlhgc/SchedulingDelegate.cpp

```diff
@@ -1157,16 +1652,42 @@ MM_SchedulingDelegate::updateGMPStats(MM_EnvironmentVLHGC *env)
 	MM_MarkVLHGCStats * concurrentMarkStats = &(persistentGMPState->_vlhgcCycleStats._concurrentMarkStats);
 
-	U_64 incrementalScanTime = (U_64) (((double) j9time_hires_delta(0, incrementalMarkStats->getScanTime(), J9PORT_TIME_DELTA_IN_MICROSECONDS)) / ((double) _extensions->gcThreadCount));
-	UDATA concurrentBytesScanned = concurrentMarkStats->_bytesScanned;
+	uint64_t incrementalScanTime = (uint64_t) (((double) j9time_hires_delta(0, incrementalMarkStats->getScanTime(), J9PORT_TIME_DELTA_IN_MICROSECONDS)) / ((double) _extensions->gcThreadCount));
+	uintptr_t concurrentBytesScanned = concurrentMarkStats->_bytesScanned;
 
-	_historicTotalIncrementalScanTimePerGMP = (U_64) ((_historicTotalIncrementalScanTimePerGMP * incrementalScanTimePerGMPHistoricWeight) + (incrementalScanTime * (1 - incrementalScanTimePerGMPHistoricWeight)));
-	_historicBytesScannedConcurrentlyPerGMP = (UDATA) ((_historicBytesScannedConcurrentlyPerGMP * bytesScannedConcurrentlyPerGMPHistoricWeight) + (concurrentBytesScanned * (1 - bytesScannedConcurrentlyPerGMPHistoricWeight)));
+	_historicTotalIncrementalScanTimePerGMP = (uint64_t) ((_historicTotalIncrementalScanTimePerGMP * incrementalScanTimePerGMPHistoricWeight) + (incrementalScanTime * (1 - incrementalScanTimePerGMPHistoricWeight)));
+	_historicBytesScannedConcurrentlyPerGMP = (uintptr_t) ((_historicBytesScannedConcurrentlyPerGMP * bytesScannedConcurrentlyPerGMPHistoricWeight) + (concurrentBytesScanned * (1 - bytesScannedConcurrentlyPerGMPHistoricWeight)));
 
 	Trc_MM_SchedulingDelegate_updateGMPStats(env->getLanguageVMThread(), _historicTotalIncrementalScanTimePerGMP, incrementalScanTime, _historicBytesScannedConcurrentlyPerGMP, concurrentBytesScanned);
 }
 
-U_64
+void 
+MM_SchedulingDelegate::updatePgcTimePrediction(MM_EnvironmentVLHGC *env)
+{
+	/* 
+	 * Create a model that passes through (minimumEdenRegions,minimumPgcTime) and (current eden size in regions, pgcTime) 
+	 * By remembering historic values of _pgcTimeIncreasePerEdenRegionFactor, it is possible to reasonably accuratly predict how long PGC will take, if eden were to change size
+	 */
+	double x1 = (double)minimumEdenRegions;
+	double y1 = (double)minimumPgcTime;
+
+	double x2 = (double)getCurrentEdenSizeInRegions(env);
+	double y2 = (double)_historicalPartialGCTime;
+
+	/* 
+	 * Calculate how closely related PGC is to eden time. The closer _pgcTimeIncreasePerEdenRegionFactor is to 1.0, the more directly changing eden size will impact pgc time.
+	 * The higher _pgcTimeIncreasePerEdenRegionFactor is from 1, the less changing eden size will affect pgc time. 
+	 * In certain edge cases where eden is very small (minimumEdenRegions in size), or pgc time is very small, skip this set of calculation, since the results will not be correct
+	 */
+	if ((x1 < x2) && (y1 < y2)) {
+		double timeDiff = y1 - y2;
+		double edenSizeRatio = ((x1 + 1.0) / (x2 + 1.0));
+		_pgcTimeIncreasePerEdenRegionFactor = pow(edenSizeRatio, (1.0/timeDiff));
```

Comment:
Just a note for reviewers - this `_pgcTimeIncreasePerEdenRegionFactor` **should not** be a weighted average.

Rationale - `_pgcTimeIncreasePerEdenRegionFactor` is not "linear" in the sense that the difference between 1.1 and 1.2, is not "the same" as 1.2 to 1.3, since this equation is finding the logarithmic fit between the 2 points. 
That being said, using a weighted average, is not really a good idea. Additionally, `_historicalPartialGCTime` is being used for this calculation, which is already a moving average, so no need to make `_pgcTimeIncreasePerEdenRegionFactor` a weighted average as well.

Reply:
This is probably ok for now, but this are still some assumed const here, which would fit for some workloads/current eden sizes and not for others. Specifically, A (assumed 1.0) here would could vary:

pow(edenSizeRatio, ( A /timeDiff));

How I would have done it (in theory anyway) is to remember last few (5-10) of (size, time) pairs  and do a linear regression. That would give you first derivative (slope of the whatever the real type of the curve is) in the _neighborhood_ of recent eden sizes. Indeed this is what we need, since we are typically doing incremental resizing (<10%), not that far from the recent past samples.

That approach would get rid of assumptions like minimumPgcTime, factor A etc, 
but it would be a lot more calculations and early estimates (during startup/rampup) would be fairly inaccurate.

Reply:
Here the `1.0`, in `pow(edenSizeRatio, ( A /timeDiff));` is a workaround to represent  nthroot. The actual formula is the following 

![image](https://user-images.githubusercontent.com/28404014/111229238-f3c3b500-85bb-11eb-93ba-c78c28fccbc6.png)

Where `z` in the image, is  `_pgcTimeIncreasePerEdenRegionFactor` in the code. That being said, I don't really see the `1.0` changing, unless we change the formula (which might be an option).

As for the suggested improvement, I agree that this would be a good refinement, and I think we had briefly spoken about doing something along those lines, but as you mention, it adds more complexity/calculations.